internal/statestore/backfill_test.go

```diff
@@ -300,4 +309,130 @@ func TestDeleteBackfill(t *testing.T) {
 	require.Error(t, err)
 	require.Equal(t, codes.Unavailable.String(), status.Convert(err).Code().String())
 	require.Contains(t, status.Convert(err).Message(), "DeleteBackfill, id: 12345, failed to connect to redis:")
+
+	// test that Backfill also deleted from last acknowledged sorted set
+	_, err = redis.Int64(conn.Do("ZSCORE", backfillLastAckTime, bfID))
+	require.Error(t, err)
+}
+
+// TestAcknowledgeBackfillLifecycle test statestore functions - AcknowledgeBackfill, GetExpiredBackfillIDs
+// and deleteExpiredBackfillID
+func TestAcknowledgeBackfillLifecycle(t *testing.T) {
+	cfg, closer := createRedis(t, false, "")
+	defer closer()
+
+	service := New(cfg)
+	require.NotNil(t, service)
+	defer service.Close()
+	ctx := utilTesting.NewContext(t)
+
+	bf1 := "mockBackfillID"
+	bf2 := "mockBackfillID2"
+	err := service.CreateBackfill(ctx, &pb.Backfill{
+		Id:         bf1,
+		Generation: 1,
+	}, nil)
+	require.NoError(t, err)
+
+	err = service.CreateBackfill(ctx, &pb.Backfill{
+		Id:         bf2,
+		Generation: 1,
+	}, nil)
+	require.NoError(t, err)
+
+	bfIDs, err := service.GetExpiredBackfillIDs(ctx)
+	require.Len(t, bfIDs, 0)
+	require.NoError(t, err)
+	err = service.AcknowledgeBackfill(ctx, bf1)
+	require.NoError(t, err)
+	err = service.AcknowledgeBackfill(ctx, bf2)
+	require.NoError(t, err)
+	// Sleep until the pending release expired and verify we still have all the tickets
+	time.Sleep(cfg.GetDuration("pendingReleaseTimeout"))
+
+	bfIDs, err = service.GetExpiredBackfillIDs(ctx)
+	require.Len(t, bfIDs, 2)
+	require.NoError(t, err)
+
+	// Acknowledge one Backfill it should be removed from GetExpired output
+	err = service.AcknowledgeBackfill(ctx, bf2)
+	require.NoError(t, err)
+	bfIDs, err = service.GetExpiredBackfillIDs(ctx)
+	require.Len(t, bfIDs, 1)
+	require.NoError(t, err)
+	require.Equal(t, bf1, bfIDs[0])
+
+	err = service.DeleteBackfill(ctx, bfIDs[0])
+	require.NoError(t, err)
+
+	bfIDs, err = service.GetExpiredBackfillIDs(ctx)
+	require.Len(t, bfIDs, 0)
+	require.NoError(t, err)
+}
+
+// TestAcknowledgeBackfill test statestore function AcknowledgeBackfill
+func TestAcknowledgeBackfill(t *testing.T) {
+	cfg, closer := createRedis(t, false, "")
+	defer closer()
+
+	curTime := time.Now()
+	service := New(cfg)
+	require.NotNil(t, service)
+	defer service.Close()
+	ctx := utilTesting.NewContext(t)
+	bf1 := "mockBackfillID"
+	err := service.CreateBackfill(ctx, &pb.Backfill{
+		Id:         bf1,
+		Generation: 1,
+	}, nil)
+	require.NoError(t, err)
+	err = service.AcknowledgeBackfill(ctx, bf1)
+	require.NoError(t, err)
+
+	// Check that Acknowledge timestamp stored valid in Redis
+	pool := GetRedisPool(cfg)
+	conn := pool.Get()
+	res, err := redis.Int64(conn.Do("ZSCORE", backfillLastAckTime, bf1))
+	require.NoError(t, err)
+	require.True(t, (time.Unix(res/1e9, res%1e9).Sub(curTime)) < time.Second)
+
+	cfg = createInvalidRedisConfig()
```

Comment:
This is a completely different test case. Please move it to another test function or add table-driven tests here.

Reply:
Created a new one